R/simulate_linreg.R

```diff
@@ -277,60 +213,41 @@ sim_sumstats <- function(RL, ngwas, beta, h2ge) {
+#' This function simulates multiple traits (phenotypes) based on genotype data, the rows of output will represent each trait, and the columns are each subjects' phenotype (trait)
+#' and heritability. It allows specifying if the heritability
 #' is total or per-SNP, optionally scales the phenotypes, and can handle residual correlations.
 #' per eQTL heritability is 0.05 to 0.07 according to https://www.nature.com/articles/ng.3506
 #'
 #' @param G Genotype matrix.
 #' @param B Matrix of effect sizes for multiple traits.
 #' @param h2g Heritability (proportion of variance explained by genetics).
 #' @param is_h2g_total Logical indicating if h2g is total (TRUE) or per-SNP (FALSE).
-#' @param max_h2g Maximum heritability allowed in the model (default is 0.8).
 #' @param residual_corr Matrix of residual correlations (NULL for independent samples).
 #' @param scale_Y Logical indicating if the phenotypes should be scaled (default is FALSE).
 #' @return A list containing the simulated phenotypes (`Y`) and residual variance (`residual_var`).
 #' @importFrom MASS mvrnorm
 #' @export
 #'
 #' @examples
-#' G <- matrix(rnorm(1000), nrow = 100, ncol = 10) # Random genotype matrix
+#' G = matrix(rbinom(1000, 2, 0.5), nrow = 1000, ncol = 50) 
 #' # Simulating effect sizes for two traits
-#' B <- sapply(1:2, function(i) sim_beta(RL, ncausal, eqtl_h2)) 
-#' h2g <- 0.5                                      # Heritability
-#' result <- sim_multi_traits(G, B, h2g)           # Simulate multiple traits
-sim_multi_traits <- function(G, B, h2g, is_h2g_total = TRUE, max_h2g = 0.8, 
-                             residual_corr = NULL, scale_Y = FALSE) {
-    if('scaled:scale' %in% names(attributes(G))) {
-        SB <- sweep(B, 2, attr(G, "scaled:scale"), FUN = "/")
-        GSB <- G %*% SB
-        Yhat <- sweep(GSB, 2, colMeans(GSB), FUN = "-")
-    } else {
-        Yhat <- G %*% B
-    }
-
-    genetic_var <- apply(Yhat, 2, var)
-
-    if (!is_h2g_total) {
-        h2g <- min(h2g * ncol(B), max_h2g)
+#' B = sim_beta(G, ncausal = 5, ntrait = 3, is_h2g_total = F, shared_pattern = "all")
+#' P = sim_multi_traits(G, B, h2g = 0.1, is_h2g_total = T, max_h2g = 1)
+sim_multi_traits = function(G, B, h2g, is_h2g_total = TRUE, max_h2g = 1){
+  P = matrix(0, nrow = ncol(B), ncol = nrow(G)) # row: traits, column: different subjects
+  mu = G %*% B 
+  for(i in 1:ncol(mu)){
+  if(is_h2g_total){
+    sigma2 = var(mu[,i]) * (1-h2g) / h2g
+    P[i,] = apply(as.matrix(mu[,i]), 1, function(mu0) {return(rnorm(1, mu0, sqrt(sigma2)))})
+  }else{
+    first_index = min(which(B[,i]!=0))
+    sigma2 =  var(G[,first_index])/h2g - var(mu[,i])
+    P[i,] = apply(as.matrix(mu[,i]), 1, function(mu0) {return(rnorm(1, mu0, sqrt(sigma2)))})
+      
     }
-
-    sigma <- sqrt(genetic_var / h2g - genetic_var)
-    if (is.null(residual_corr)) {
-        residual_corr <- diag(nrow(Yhat))
-    }
-
-    residual_var <- sweep(sweep(residual_corr, 2, sigma, "*"), 1, sigma, "*")
```

Comment:
this is supposed to be a matrix. How it works is this: U = SRS (matrix multiplication) where U is residual variance, S = diag(sigma) (not squared) is the marginal standard deviation, and R is residual correlation. When R is a diagonal matrix, it is the same as when you simulate each column of Y through a loop assuming independence, as what you do now. 

As a good exercise please do this: create a function when residual_corr is NULL, do it your way looping through all columns. If residual corr is not NULL do it the multivariate approach. What you will be testing is when you set residual corr to NULL compared to when you set it to a diag matrix, the output should be identical. Please make it a unit test using `testthat`